library/src/schemas/variant/variant.ts

```diff
@@ -6,13 +6,15 @@ import type {
   Output,
 } from '../../types/index.ts';
 import { getSchemaIssues, getOutput, getIssues } from '../../utils/index.ts';
-import type { ObjectSchema } from '../object/index.ts';
+import type { Literal } from '../literal/index.ts';
+import type { LiteralSchema } from '../literal/literal.ts';
+import type { ObjectPathItem, ObjectSchema } from '../object/index.ts';
 
 /**
  * Variant option type.
  */
 export type VariantOption<TKey extends string> =
-  | ObjectSchema<Record<TKey, BaseSchema>, any>
+  | ObjectSchema<Record<TKey, LiteralSchema<Literal>>, any>
```

Comment:
Yes, as discussed [here](https://github.com/fabian-hiller/valibot/issues/235#issuecomment-1791837619). Why do you think it would make `variant` useless? I have a hard time finding real-world use-cases with anything else than literals. Zod also doesn’t support much more, see [here](https://github.com/colinhacks/zod/blob/05d33f22b1001f29d8c3d2b1ded16a089a535072/src/types.ts#L2886-L2909). Although Zod supports some special cases like literals that are wrapped in `ZodLazy` or `ZodEffects` as well as `null`, `undefined` and `enums`.

Reply:
Ah now I get your point. I need to take a closer look at `VariantOption` and `ObjectSchema`. Because from just reading the code I come to the same conclusion as you. But for some reason the example you provided works fine. See also the unit tests I’ve added in this PR, the `val` property uses non-literal schemas and there is no typescript compilation error... 🤔

Reply:
Yes, I had thought of that and wondered about it. What is the reason why we can only accept literals as discriminators? Presumably you have to access the `.literal` for the error info, right?

Reply:
> What is the reason why we can only accept literals as discriminators? Presumably you have to access the `.literal` for the error info, right?

Exactly. An alternative solution would be to just pass the schema objects of every option to the issue requirements instead of limiting the schemas to literal schemas. In this case the consumer needs to decide what to do with the requirements and could extract the `.literal` value in case the schema object is a literal schema.

Reply:
I will think about it. However, I probably won't be able to work on this PR for another three to four weeks.